attribution/experiment_logger.py

```diff
@@ -173,14 +176,126 @@ def log_perturbation(
+
+            html_str += f" -> <b>{output}</b>"
+            html_str += "</div>"
+
+            # Display
+
+            if get_ipython() and "IPKernelApp" in get_ipython().config:
+                from IPython.display import display
+
+                display(HTML(html_str))
+            else:
+                self.pretty_print(df)
+
+    def print_text_attribution_matrix(self, exp_id: int = -1):
+        if exp_id == -1:
+            exp_id = self.df_experiments["exp_id"].max()
+
+        matrix = self.get_attribution_matrix(exp_id)
+
+        input_tokens = [" ".join(x.split(" ")[:-1]) for x in matrix.index]
+        token_dict = {f"token_{i+1}": t for i, t in enumerate(input_tokens)}
+
+        for oi, output_token in enumerate(matrix.columns):
+            prev_output_str = "".join([" ".join(ot.split(" ")[:-1]) for ot in matrix.columns[:oi]])
+            following_output_str = "".join(
+                [" ".join(ot.split(" ")[:-1]) for ot in matrix.columns[oi + 1 :]]
+            )
+            attr_scores = matrix[output_token].tolist()
+            vmax = max(abs(min(attr_scores)), abs(max(attr_scores)))
+            vmin = -vmax
+
+            score_dict = {f"token_{i+1}": score for i, score in enumerate(attr_scores)}
+
+            df = pd.DataFrame([token_dict, score_dict], index=["token", "attr_score"])
+
+            # Generating HTML
+            html_str = '<div style="font-family: monospace;">'
+            for col in df.columns:
+                token = df[col]["token"]
+                score = df[col]["attr_score"]
+                color = self.score_to_color(score, vmin, vmax)
+                html_str += f'<span style="text-decoration: underline; text-decoration-color: {color}; text-decoration-thickness: 4px;">{token}</span>'
+
+            clean_output_token = " ".join(output_token.split(" ")[:-1])
+            html_str += f" -> {prev_output_str}<b>{clean_output_token}</b>{following_output_str}"
+            html_str += "</div>"
+
+            # Display
+
+            if get_ipython() and "IPKernelApp" in get_ipython().config:
+                from IPython.display import display
+
+                display(HTML(html_str))
+            else:
+                self.pretty_print(df)
+
+    def print_total_attribution(
+        self, exp_id: Optional[int] = None, score_agg: Literal["mean", "last"] = "mean"
+    ):
+        totals = []
+        if exp_id == -1:
+            exp_id = self.df_experiments["exp_id"].max()
+
+        token_attrs_df = (
+            self.df_input_token_attribution.groupby(["exp_id", "attribution_strategy"])
+            if exp_id is None
+            else self.df_input_token_attribution[
+                self.df_input_token_attribution["exp_id"] == exp_id
+            ].groupby(["exp_id", "attribution_strategy"])
+        )
+
+        for (exp_id, attr_strat), exp_data in token_attrs_df:
```

Comment:
Small nitpick, but can we use `attr_strategy` for clarity?

Reply:
sure